localization_manager/src/LocalizationTransitionTable.cpp

```diff
@@ -0,0 +1,230 @@
+      logDebugSignal(signal);
+      break;
+  }
+}
+void LocalizationTransitionTable::signalWhenOPERATIONAL(LocalizationSignal signal)
+{
+  switch (signal)
+  {
+    case LocalizationSignal::INITIAL_POSE:
+      setAndLogState(LocalizationState::INITIALIZING, signal);
+      break;
+    case LocalizationSignal::POOR_NDT_FREQ_OR_FITNESS_SCORE:
+      setAndLogState(LocalizationState::DEGRADED, signal);
+      break;
+    case LocalizationSignal::LIDAR_SENSOR_FAILURE:
+      if (mode_ == LocalizerMode::AUTO)
+      {
+        setAndLogState(LocalizationState::DEGRADED_NO_LIDAR_FIX, signal);
+      }
+      else
+      {
+        setAndLogState(LocalizationState::AWAIT_MANUAL_INITIALIZATION, signal);
+      }
+      break;
+    case LocalizationSignal::UNUSABLE_NDT_FREQ_OR_FITNESS_SCORE:
+      if (mode_ == LocalizerMode::AUTO)
+      {
+        setAndLogState(LocalizationState::DEGRADED_NO_LIDAR_FIX, signal);
+      }
+      else
+      {
+        setAndLogState(LocalizationState::AWAIT_MANUAL_INITIALIZATION, signal);
+      }
+      break;
+    default:
+      logDebugSignal(signal);
+      break;
+  }
+}
+void LocalizationTransitionTable::signalWhenDEGRADED(LocalizationSignal signal)
+{
+  switch (signal)
+  {
+    case LocalizationSignal::INITIAL_POSE:
+      setAndLogState(LocalizationState::INITIALIZING, signal);
+      break;
+    case LocalizationSignal::GOOD_NDT_FREQ_AND_FITNESS_SCORE:
+      setAndLogState(LocalizationState::OPERATIONAL, signal);
+      break;
+    case LocalizationSignal::LIDAR_SENSOR_FAILURE:
+      if (mode_ == LocalizerMode::AUTO)
+      {
+        setAndLogState(LocalizationState::DEGRADED_NO_LIDAR_FIX, signal);
+      }
+      else
+      {
+        setAndLogState(LocalizationState::AWAIT_MANUAL_INITIALIZATION, signal);
+      }
+      break;
+    case LocalizationSignal::UNUSABLE_NDT_FREQ_OR_FITNESS_SCORE:
+      if (mode_ == LocalizerMode::AUTO)
+      {
+        setAndLogState(LocalizationState::DEGRADED_NO_LIDAR_FIX, signal);
+      }
+      else
+      {
+        setAndLogState(LocalizationState::AWAIT_MANUAL_INITIALIZATION, signal);
+      }
+      break;
+    default:
+      logDebugSignal(signal);
+      break;
+  }
+}
+void LocalizationTransitionTable::signalWhenDEGRADED_NO_LIDAR_FIX(LocalizationSignal signal)
+{
+  switch (signal)
+  {
+    case LocalizationSignal::INITIAL_POSE:
+      if (mode_ != LocalizerMode::GNSS)
+      {
+        setAndLogState(LocalizationState::INITIALIZING, signal);
+      }
+      break;
+    case LocalizationSignal::TIMEOUT:
+      if (mode_ != LocalizerMode::GNSS)
+      {
+        setAndLogState(LocalizationState::AWAIT_MANUAL_INITIALIZATION, signal);
+      }
+      break;
+    default:
+      logDebugSignal(signal);
+      break;
+  }
+}
+void LocalizationTransitionTable::signalWhenAWAIT_MANUAL_INITIALIZATION(LocalizationSignal signal)
+{
+  switch (signal)
+  {
+    case LocalizationSignal::INITIAL_POSE:
+      setAndLogState(LocalizationState::INITIALIZING, signal);
```

Comment:
On the state diagram, it seems like we need to check the mode here?

Reply:
It's not actually possible to enter AWAIT_MANUAL_INITIALIZATION while in GNSS mode, so the check is implicit.